Notify on feature flag changes

emord opened the pull request:
@esoergel @orangejenny we talked about this last week. This notifies the team on any deprecated flag changes and you can also specify yourself on any you want to track yourself.

Also considered a thank you email for removed entries. maybe some positive reinforcement would be good?

emord committed:
Warn us about deprecated toggle usage

emord committed:
Add emord to ucr toggle

esoergel commented on corehq/apps/toggle_ui/views.py:
```diff
@@ -194,6 +197,26 @@ def post(self, request, *args, **kwargs):
+            else:
+                _assert = soft_assert(send_to_ops=is_deprecated_toggle)
+
+            _assert(False, subject)
```
You could stick this stuff in a private method to keep the `post` method relatively small.

emord replied:
https://github.com/dimagi/commcare-hq/pull/21273/commits/4a56ebeba9e34b390f4a88e6c71641fa6aec98c7

orangejenny commented on corehq/apps/toggle_ui/views.py:
```diff
@@ -194,6 +197,26 @@ def post(self, request, *args, **kwargs):
         changed_entries = previously_enabled ^ currently_enabled  # ^ means XOR
+
+        added_entries = currently_enabled - previously_enabled
+        is_deprecated_toggle = (self.static_toggle.tag == TAG_DEPRECATED)
```
I could imagine this also applying to TAG_CUSTOM and TAG_INTERNAL - anecdotally, I feel like a number of the issues we have are when a flag built for a project like ICDS starts getting used on solutions projects. Or would that lead to too many emails?

orangejenny commented:
I like the ability to subscribe to individual flags. You're going to announce this to the dev team, yes?

orangejenny commented:
> Also considered a thank you email for removed entries.

I am mildly curious if anyone ever turns *off* a  feature flag.

emord committed:
Move notification to private method

emord committed:
Remove incorrect translation syntax

Just removed it from translations entirely since it's an admin
interface, would add noise to transifex

emord committed:
Add custom and internal toggle notifications

esoergel commented on corehq/apps/toggle_ui/views.py:
```diff
@@ -192,7 +192,7 @@ def post(self, request, *args, **kwargs):
                 _clear_caches_for_dynamic_toggle(self.toggle_meta())
 
         elif save_randomness:
-            messages.error(request, _("The randomness value {} must be between 0 and 1".format(randomness)))
```
:+1:

emord commented:
@dimagi/team-commcare-hq You can now subscribe to feature flag additions. 

This PR allows you to add your username to a feature flag and it will email you whenever someone adds a domain or user to it. This way you can proactively track any usage for a feature flag that is not intended for general availability.

In addition if anyone adds someone to an internal, custom or deprecated feature flag, it will send an email to the team.
